File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-go/Mailstandard/provider_extension_catalog.go
package Mailstandard

type MailProviderExtensionCatalogEntry struct {
    ExtensionKey string
    ProviderKey  string
    DisplayName  string
    Surface      string
    Access       string
    Status       string
}

type MailProviderExtensionCatalog struct{}

var mail_PROVIDER_EXTENSION_SURFACES = []string{
    "control-plane",
    "runtime-bridge",
    "cross-surface",
}

var mail_PROVIDER_EXTENSION_ACCESSES = []string{
    "unwrap-only",
    "extension-object",
}

var mail_PROVIDER_EXTENSION_STATUSES = []string{
    "reference-baseline",
    "reserved",
}

var mail_PROVIDER_EXTENSION_CATALOG = []MailProviderExtensionCatalogEntry{
    {ExtensionKey: "smtp.transport", ProviderKey: "smtp", DisplayName: "SMTP Transport", Surface: "runtime-bridge", Access: "unwrap-only", Status: "reserved"},
    {ExtensionKey: "imap.sync", ProviderKey: "imap", DisplayName: "IMAP Sync", Surface: "runtime-bridge", Access: "unwrap-only", Status: "reserved"},
}

func GetMailProviderExtensionCatalog() []MailProviderExtensionCatalogEntry {
    return append([]MailProviderExtensionCatalogEntry(nil), mail_PROVIDER_EXTENSION_CATALOG...)
}

func GetMailProviderExtensionDescriptor(extensionKey string) *MailProviderExtensionCatalogEntry {
    for index := range mail_PROVIDER_EXTENSION_CATALOG {
        if mail_PROVIDER_EXTENSION_CATALOG[index].ExtensionKey == extensionKey {
            return &mail_PROVIDER_EXTENSION_CATALOG[index]
        }
    }

    return nil
}

func GetMailProviderExtensionsForProvider(providerKey string) []MailProviderExtensionCatalogEntry {
    entries := make([]MailProviderExtensionCatalogEntry, 0)
    for _, entry := range mail_PROVIDER_EXTENSION_CATALOG {
        if entry.ProviderKey == providerKey {
            entries = append(entries, entry)
        }
    }

    return entries
}

func GetMailProviderExtensions(extensionKeys []string) []MailProviderExtensionCatalogEntry {
    entries := make([]MailProviderExtensionCatalogEntry, 0)
    for _, extensionKey := range extensionKeys {
        entry := GetMailProviderExtensionDescriptor(extensionKey)
        if entry != nil {
            entries = append(entries, *entry)
        }
    }

    return entries
}

func HasMailProviderExtension(extensionKeys []string, extensionKey string) bool {
    if GetMailProviderExtensionDescriptor(extensionKey) == nil {
        return false
    }

    for _, value := range extensionKeys {
        if value == extensionKey {
            return true
        }
    }

    return false
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-rust/src/provider_extension_catalog.rs
#[derive(Clone, Copy)]
#[allow(non_snake_case)]
pub struct MailProviderExtensionCatalogEntry {
    pub extensionKey: &'static str,
    pub providerKey: &'static str,
    pub displayName: &'static str,
    pub surface: &'static str,
    pub access: &'static str,
    pub status: &'static str,
}

pub struct MailProviderExtensionCatalog;

pub const mail_PROVIDER_EXTENSION_SURFACES: [&str; 3] = [
    "control-plane",
    "runtime-bridge",
    "cross-surface",
];

pub const mail_PROVIDER_EXTENSION_ACCESSES: [&str; 2] = [
    "unwrap-only",
    "extension-object",
];

pub const mail_PROVIDER_EXTENSION_STATUSES: [&str; 2] = [
    "reference-baseline",
    "reserved",
];

pub const mail_PROVIDER_EXTENSION_CATALOG: [MailProviderExtensionCatalogEntry; 2] = [
    MailProviderExtensionCatalogEntry { extensionKey: "smtp.transport", providerKey: "smtp", displayName: "SMTP Transport", surface: "runtime-bridge", access: "unwrap-only", status: "reserved" },
    MailProviderExtensionCatalogEntry { extensionKey: "imap.sync", providerKey: "imap", displayName: "IMAP Sync", surface: "runtime-bridge", access: "unwrap-only", status: "reserved" },
];

pub fn get_mail_provider_extension_catalog() -> &'static [MailProviderExtensionCatalogEntry] {
    &mail_PROVIDER_EXTENSION_CATALOG
}

pub fn get_mail_provider_extension_descriptor(
    extension_key: &str,
) -> Option<&'static MailProviderExtensionCatalogEntry> {
    mail_PROVIDER_EXTENSION_CATALOG
        .iter()
        .find(|entry| entry.extensionKey == extension_key)
}

pub fn get_mail_provider_extensions_for_provider(
    provider_key: &str,
) -> Vec<MailProviderExtensionCatalogEntry> {
    mail_PROVIDER_EXTENSION_CATALOG
        .iter()
        .filter(|entry| entry.providerKey == provider_key)
        .copied()
        .collect()
}

pub fn get_mail_provider_extensions(
    extension_keys: &[&str],
) -> Vec<MailProviderExtensionCatalogEntry> {
    extension_keys
        .iter()
        .filter_map(|extension_key| get_mail_provider_extension_descriptor(extension_key).copied())
        .collect()
}

pub fn has_mail_provider_extension(extension_keys: &[&str], extension_key: &str) -> bool {
    extension_keys.iter().any(|value| *value == extension_key)
        && get_mail_provider_extension_descriptor(extension_key).is_some()
}
